pandas/core/arrays/datetimes.py

```diff
@@ -1647,6 +1649,19 @@ def maybe_convert_dtype(data, copy):
         raise TypeError("Passing PeriodDtype data is invalid.  "
                         "Use `data.to_timestamp()` instead")
 
+    elif is_categorical_dtype(data):
+        # TODO: cases where we need to do another pass through this func,
+        #  e.g. the categories are timedelta64s
+        if isna(data).any():
+            # slow-path
+            data = list(data)
+            data = np.array(data, dtype=np.object_)
+            copy = False
+        else:
+            # TODO: does this always make a copy?  If so, set copy=False
+            # GH#18664 preserve tz in going DTI->Categorical->DTI
+            data = data.categories[data.codes]
```

Comment:
shouldn't this be take?

Reply:
I know this works, don't know if there is a more efficient implementation.